pkg/client/unit_test.go

```diff
@@ -0,0 +1,58 @@
+package client
+
+import (
+	"testing"
+
+	"github.com/elastic/elastic-agent-client/v7/pkg/proto"
+	"github.com/stretchr/testify/require"
+	"google.golang.org/protobuf/types/known/structpb"
+)
+
+var defaultTest = Unit{
+	exp:       UnitStateHealthy,
+	logLevel:  UnitLogLevelDebug,
+	configIdx: 1,
+	config:    &proto.UnitExpectedConfig{},
+}
+
+func TestUnitUpdateWithSameMap(t *testing.T) {
+	defaultTest.configIdx = 1
+	sameMap := map[string]interface{}{"username": "test"}
+	pbStruct, err := structpb.NewStruct(sameMap)
+	require.NoError(t, err)
+	defaultTest.config.Source = pbStruct
+
+	pbStructNew, err := structpb.NewStruct(sameMap)
+	newUnit := &proto.UnitExpectedConfig{
+		Source: pbStructNew,
+	}
+	// This should return false, as the two underlying maps in `source` are the same
```

Comment:
If you add a call to `proto.Marshal` it would fail if you fell back to using `reflect.DeepEqual`. I think it's worth adding this to ensure the implementation doesn't change later.

```diff
diff --git a/pkg/client/unit_test.go b/pkg/client/unit_test.go
index e4539a6..610cfc5 100644
--- a/pkg/client/unit_test.go
+++ b/pkg/client/unit_test.go
@@ -5,6 +5,7 @@ import (

        "github.com/elastic/elastic-agent-client/v7/pkg/proto"
        "github.com/stretchr/testify/require"
+       gproto "google.golang.org/protobuf/proto"
        "google.golang.org/protobuf/types/known/structpb"
 )

@@ -26,6 +27,11 @@ func TestUnitUpdateWithSameMap(t *testing.T) {
        newUnit := &proto.UnitExpectedConfig{
                Source: pbStructNew,
        }
+
+       // Marshal the message to populate the size cache and ensure the internal proto files are populated.
+       _, err = gproto.Marshal(newUnit)
+       require.NoError(t, err)
+
        // This should return false, as the two underlying maps in `source` are the same
        result := defaultTest.updateState(UnitStateHealthy, UnitLogLevelDebug, newUnit, 2)
        require.False(t, result)
(END)
```

Reply:
(I confirmed this test does fail if you switch the comparison to use `reflect.DeepEqual` instead of `proto.Equal`)

Reply:
Ahh, didn't even think of that.

Reply:
I have had this bug before :)